File: fons/nanus/ast.ts
/**
 * NANUS - Minimal Faber Compiler
 *
 * AST type definitions for the subset of Faber needed to compile rivus.
 * This is intentionally minimal: 84 features, not 139.
 *
 * Design: Discriminated unions with 'tag' field. No base classes, no visitors.
 * Extend by adding variants to the unions and cases to the switches.
 */

// Source location for error reporting
export interface Locus {
    linea: number;
    columna: number;
    index: number;
}

// Tokens produced by lexer
export interface Token {
    tag: TokenTag;
    valor: string;
    locus: Locus;
}

export type TokenTag =
    | 'EOF'
    | 'Newline'
    | 'Identifier'
    | 'Numerus'
    | 'Textus'
    | 'Operator'
    | 'Punctuator'
    | 'Keyword'
    | 'Comment';

// Type annotations
export type Typus =
    | { tag: 'Nomen'; nomen: string }
    | { tag: 'Nullabilis'; inner: Typus }
    | { tag: 'Genericus'; nomen: string; args: Typus[] }
    | { tag: 'Functio'; params: Typus[]; returns: Typus }
    | { tag: 'Unio'; members: Typus[] }
    | { tag: 'Litteralis'; valor: string };

// Expressions
export type Expr =
    | { tag: 'Nomen'; locus: Locus; valor: string }
    | { tag: 'Ego'; locus: Locus }
    | { tag: 'Littera'; locus: Locus; species: LitteraSpecies; valor: string }
    | { tag: 'Binaria'; locus: Locus; signum: string; sin: Expr; dex: Expr }
    | { tag: 'Unaria'; locus: Locus; signum: string; arg: Expr }
    | { tag: 'Assignatio'; locus: Locus; signum: string; sin: Expr; dex: Expr }
    | { tag: 'Condicio'; locus: Locus; cond: Expr; cons: Expr; alt: Expr }
    | { tag: 'Vocatio'; locus: Locus; callee: Expr; args: Expr[] }
    | { tag: 'Membrum'; locus: Locus; obj: Expr; prop: Expr; computed: boolean; nonNull: boolean }
    | { tag: 'Series'; locus: Locus; elementa: Expr[] }
    | { tag: 'Obiectum'; locus: Locus; props: ObiectumProp[] }
    | { tag: 'Clausura'; locus: Locus; params: Param[]; corpus: Expr | Stmt }
    | { tag: 'Novum'; locus: Locus; callee: Expr; args: Expr[]; init: Expr | null }
    | { tag: 'Cede'; locus: Locus; arg: Expr }
    | { tag: 'Qua'; locus: Locus; expr: Expr; typus: Typus }
    | { tag: 'Innatum'; locus: Locus; expr: Expr; typus: Typus }
    | { tag: 'Finge'; locus: Locus; variant: string; campi: ObiectumProp[]; typus: Typus | null }
    | { tag: 'Scriptum'; locus: Locus; template: string; args: Expr[] }
    | { tag: 'Ambitus'; locus: Locus; start: Expr; end: Expr; inclusive: boolean };

export type LitteraSpecies = 'Numerus' | 'Fractus' | 'Textus' | 'Verum' | 'Falsum' | 'Nihil';

export interface ObiectumProp {
    locus: Locus;
    key: Expr;
    valor: Expr;
    shorthand: boolean;
    computed: boolean;
}

// Statements
export type Stmt =
    | { tag: 'Massa'; locus: Locus; corpus: Stmt[] }
    | { tag: 'Expressia'; locus: Locus; expr: Expr }
    | { tag: 'Varia'; locus: Locus; species: VariaSpecies; nomen: string; typus: Typus | null; valor: Expr | null; publica: boolean; externa: boolean }
    | { tag: 'Functio'; locus: Locus; nomen: string; params: Param[]; typusReditus: Typus | null; corpus: Stmt | null; asynca: boolean; publica: boolean; generics: string[] }
    | { tag: 'Genus'; locus: Locus; nomen: string; campi: CampusDecl[]; methodi: Stmt[]; implet: string[]; generics: string[]; publica: boolean }
    | { tag: 'Pactum'; locus: Locus; nomen: string; methodi: PactumMethodus[]; generics: string[]; publica: boolean }
    | { tag: 'Ordo'; locus: Locus; nomen: string; membra: OrdoMembrum[]; publica: boolean }
    | { tag: 'Discretio'; locus: Locus; nomen: string; variantes: VariansDecl[]; generics: string[]; publica: boolean }
    | { tag: 'Importa'; locus: Locus; fons: string; specs: ImportSpec[]; totum: boolean; alias: string | null }
    | { tag: 'Si'; locus: Locus; cond: Expr; cons: Stmt; alt: Stmt | null }
    | { tag: 'Dum'; locus: Locus; cond: Expr; corpus: Stmt }
    | { tag: 'FacDum'; locus: Locus; corpus: Stmt; cond: Expr }
    | { tag: 'Iteratio'; locus: Locus; species: 'Ex' | 'De'; binding: string; iter: Expr; corpus: Stmt; asynca: boolean }
    | { tag: 'Elige'; locus: Locus; discrim: Expr; casus: EligeCasus[]; default_: Stmt | null }
    | { tag: 'Discerne'; locus: Locus; discrim: Expr[]; casus: DiscerneCasus[] }
    | { tag: 'Custodi'; locus: Locus; clausulae: CustodiClausula[] }
    | { tag: 'Tempta'; locus: Locus; corpus: Stmt; cape: CapeClausula | null; demum: Stmt | null }
    | { tag: 'Redde'; locus: Locus; valor: Expr | null }
    | { tag: 'Iace'; locus: Locus; arg: Expr; fatale: boolean }
    | { tag: 'Scribe'; locus: Locus; gradus: 'Scribe' | 'Vide' | 'Mone'; args: Expr[] }
    | { tag: 'Adfirma'; locus: Locus; cond: Expr; msg: Expr | null }
    | { tag: 'Rumpe'; locus: Locus }
    | { tag: 'Perge'; locus: Locus }
    | { tag: 'Incipit'; locus: Locus; corpus: Stmt; asynca: boolean }
    | { tag: 'Probandum'; locus: Locus; nomen: string; corpus: Stmt[] }
    | { tag: 'Proba'; locus: Locus; nomen: string; corpus: Stmt };

export type VariaSpecies = 'Varia' | 'Fixum' | 'Figendum';

export interface Param {
    locus: Locus;
    nomen: string;
    typus: Typus | null;
    default_: Expr | null;
    rest: boolean;
}

export interface CampusDecl {
    locus: Locus;
    nomen: string;
    typus: Typus;
    valor: Expr | null;
    visibilitas: 'Publica' | 'Privata' | 'Protecta';
}

export interface PactumMethodus {
    locus: Locus;
    nomen: string;
    params: Param[];
    typusReditus: Typus | null;
    asynca: boolean;
}

export interface OrdoMembrum {
    locus: Locus;
    nomen: string;
    valor: string | null;
}

export interface VariansDecl {
    locus: Locus;
    nomen: string;
    campi: { nomen: string; typus: Typus }[];
}

export interface ImportSpec {
    locus: Locus;
    imported: string;
    local: string;
}

export interface EligeCasus {
    locus: Locus;
    cond: Expr;
    corpus: Stmt;
}

export interface DiscerneCasus {
    locus: Locus;
    patterns: VariansPattern[];
    corpus: Stmt;
}

export interface VariansPattern {
    locus: Locus;
    variant: string;
    bindings: string[];
    alias: string | null;
    wildcard: boolean;
}

export interface CustodiClausula {
    locus: Locus;
    cond: Expr;
    corpus: Stmt;
}

export interface CapeClausula {
    locus: Locus;
    param: string;
    corpus: Stmt;
}

// Top-level compilation unit
export interface Modulus {
    locus: Locus;
    corpus: Stmt[];
}

File: fons/nanus-rs/subsidia/src/ast.rs
use serde::{Deserialize, Serialize};

/// Source location for error reporting.
#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize)]
pub struct Locus {
    pub linea: i32,
    pub columna: i32,
    pub index: i32,
}

/// Token produced by lexer.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Token {
    pub tag: String,
    pub valor: String,
    pub locus: Locus,
}

/// Token tag constants.
pub const TOKEN_EOF: &str = "EOF";
pub const TOKEN_NEWLINE: &str = "Newline";
pub const TOKEN_IDENTIFIER: &str = "Identifier";
pub const TOKEN_NUMERUS: &str = "Numerus";
pub const TOKEN_TEXTUS: &str = "Textus";
pub const TOKEN_OPERATOR: &str = "Operator";
pub const TOKEN_PUNCTUATOR: &str = "Punctuator";
pub const TOKEN_KEYWORD: &str = "Keyword";
pub const TOKEN_COMMENT: &str = "Comment";

/// Literal species for ExprLittera.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum LitteraSpecies {
    Numerus,
    Fractus,
    Textus,
    Verum,
    Falsum,
    Nihil,
}

impl Default for LitteraSpecies {
    fn default() -> Self {
        LitteraSpecies::Nihil
    }
}

/// Variable declaration species.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum VariaSpecies {
    Varia,
    Fixum,
}

impl Default for VariaSpecies {
    fn default() -> Self {
        VariaSpecies::Varia
    }
}

/// Type annotations in AST.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "tag")]
pub enum Typus {
    #[serde(rename = "Nomen")]
    Nomen { nomen: String },

    #[serde(rename = "Nullabilis")]
    Nullabilis { inner: Box<Typus> },

    #[serde(rename = "Genericus")]
    Genericus { nomen: String, args: Vec<Typus> },

    #[serde(rename = "Functio")]
    Functio {
        params: Vec<Typus>,
        returns: Option<Box<Typus>>,
    },

    #[serde(rename = "Unio")]
    Unio { members: Vec<Typus> },

    #[serde(rename = "Litteralis")]
    Litteralis { valor: String },
}

/// Expressions in AST.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "tag")]
pub enum Expr {
    #[serde(rename = "Nomen")]
    Nomen { locus: Locus, valor: String },

    #[serde(rename = "Ego")]
    Ego { locus: Locus },

    #[serde(rename = "Littera")]
    Littera {
        locus: Locus,
        species: LitteraSpecies,
        valor: String,
    },

    #[serde(rename = "Binaria")]
    Binaria {
        locus: Locus,
        signum: String,
        sin: Box<Expr>,
        dex: Box<Expr>,
    },

    #[serde(rename = "Unaria")]
    Unaria {
        locus: Locus,
        signum: String,
        arg: Box<Expr>,
    },

    #[serde(rename = "Assignatio")]
    Assignatio {
        locus: Locus,
        signum: String,
        sin: Box<Expr>,
        dex: Box<Expr>,
    },

    #[serde(rename = "Condicio")]
    Condicio {
        locus: Locus,
        cond: Box<Expr>,
        cons: Box<Expr>,
        alt: Box<Expr>,
    },

    #[serde(rename = "Vocatio")]
    Vocatio {
        locus: Locus,
        callee: Box<Expr>,
        args: Vec<Expr>,
    },

    #[serde(rename = "Membrum")]
    Membrum {
        locus: Locus,
        obj: Box<Expr>,
        prop: Box<Expr>,
        computed: bool,
        #[serde(rename = "nonNull")]
        non_null: bool,
    },

    #[serde(rename = "Series")]
    Series { locus: Locus, elementa: Vec<Expr> },

    #[serde(rename = "Obiectum")]
    Obiectum {
        locus: Locus,
        props: Vec<ObiectumProp>,
    },

    #[serde(rename = "Clausura")]
    Clausura {
        locus: Locus,
        params: Vec<Param>,
        corpus: ClausuraCorpus,
    },

    #[serde(rename = "Novum")]
    Novum {
        locus: Locus,
        callee: Box<Expr>,
        args: Vec<Expr>,
        init: Option<Box<Expr>>,
    },

    #[serde(rename = "Cede")]
    Cede { locus: Locus, arg: Box<Expr> },

    #[serde(rename = "Qua")]
    Qua {
        locus: Locus,
        expr: Box<Expr>,
        typus: Typus,
    },

    #[serde(rename = "Innatum")]
    Innatum {
        locus: Locus,
        expr: Box<Expr>,
        typus: Typus,
    },

    #[serde(rename = "Conversio")]
    Conversio {
        locus: Locus,
        expr: Box<Expr>,
        species: String,
        fallback: Option<Box<Expr>>,
    },

    #[serde(rename = "PostfixNovum")]
    PostfixNovum {
        locus: Locus,
        expr: Box<Expr>,
        typus: Typus,
    },

    #[serde(rename = "Finge")]
    Finge {
        locus: Locus,
        variant: String,
        campi: Vec<ObiectumProp>,
        typus: Option<Typus>,
    },

    #[serde(rename = "Scriptum")]
    Scriptum {
        locus: Locus,
        template: String,
        args: Vec<Expr>,
    },

    #[serde(rename = "Ambitus")]
    Ambitus {
        locus: Locus,
        start: Box<Expr>,
        end: Box<Expr>,
        inclusive: bool,
    },
}

impl Expr {
    pub fn locus(&self) -> Locus {
        match self {
            Expr::Nomen { locus, .. } => *locus,
            Expr::Ego { locus } => *locus,
            Expr::Littera { locus, .. } => *locus,
            Expr::Binaria { locus, .. } => *locus,
            Expr::Unaria { locus, .. } => *locus,
            Expr::Assignatio { locus, .. } => *locus,
            Expr::Condicio { locus, .. } => *locus,
            Expr::Vocatio { locus, .. } => *locus,
            Expr::Membrum { locus, .. } => *locus,
            Expr::Series { locus, .. } => *locus,
            Expr::Obiectum { locus, .. } => *locus,
            Expr::Clausura { locus, .. } => *locus,
            Expr::Novum { locus, .. } => *locus,
            Expr::Cede { locus, .. } => *locus,
            Expr::Qua { locus, .. } => *locus,
            Expr::Innatum { locus, .. } => *locus,
            Expr::Conversio { locus, .. } => *locus,
            Expr::PostfixNovum { locus, .. } => *locus,
            Expr::Finge { locus, .. } => *locus,
            Expr::Scriptum { locus, .. } => *locus,
            Expr::Ambitus { locus, .. } => *locus,
        }
    }
}

/// Clausura body can be either a statement or an expression.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ClausuraCorpus {
    Stmt(Box<Stmt>),
    Expr(Box<Expr>),
}

/// Object literal property.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ObiectumProp {
    pub locus: Locus,
    pub key: Expr,
    pub valor: Expr,
    pub shorthand: bool,
    pub computed: bool,
}

/// Statements in AST.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "tag")]
pub enum Stmt {
    #[serde(rename = "Massa")]
    Massa { locus: Locus, corpus: Vec<Stmt> },

    #[serde(rename = "Expressia")]
    Expressia { locus: Locus, expr: Expr },

    #[serde(rename = "Varia")]
    Varia {
        locus: Locus,
        species: VariaSpecies,
        nomen: String,
        typus: Option<Typus>,
        valor: Option<Expr>,
        publica: bool,
        externa: bool,
    },

    #[serde(rename = "Functio")]
    Functio {
        locus: Locus,
        nomen: String,
        params: Vec<Param>,
        #[serde(rename = "typusReditus")]
        typus_reditus: Option<Typus>,
        corpus: Option<Box<Stmt>>,
        asynca: bool,
        publica: bool,
        generics: Vec<String>,
        externa: bool,
    },

    #[serde(rename = "Genus")]
    Genus {
        locus: Locus,
        nomen: String,
        campi: Vec<CampusDecl>,
        methodi: Vec<Stmt>,
        implet: Vec<String>,
        generics: Vec<String>,
        publica: bool,
        abstractus: bool,
    },

    #[serde(rename = "Pactum")]
    Pactum {
        locus: Locus,
        nomen: String,
        methodi: Vec<PactumMethodus>,
        generics: Vec<String>,
        publica: bool,
    },

    #[serde(rename = "Ordo")]
    Ordo {
        locus: Locus,
        nomen: String,
        membra: Vec<OrdoMembrum>,
        publica: bool,
    },

    #[serde(rename = "Discretio")]
    Discretio {
        locus: Locus,
        nomen: String,
        variantes: Vec<VariansDecl>,
        generics: Vec<String>,
        publica: bool,
    },

    #[serde(rename = "TypusAlias")]
    TypusAlias {
        locus: Locus,
        nomen: String,
        typus: Typus,
        publica: bool,
    },

    #[serde(rename = "In")]
    In {
        locus: Locus,
        expr: Expr,
        corpus: Box<Stmt>,
    },

    #[serde(rename = "Importa")]
    Importa {
        locus: Locus,
        fons: String,
        imported: Option<String>,
        local: String,
        totum: bool,
        publica: bool,
    },

    #[serde(rename = "Si")]
    Si {
        locus: Locus,
        cond: Expr,
        cons: Box<Stmt>,
        alt: Option<Box<Stmt>>,
    },

    #[serde(rename = "Dum")]
    Dum {
        locus: Locus,
        cond: Expr,
        corpus: Box<Stmt>,
    },

    #[serde(rename = "FacDum")]
    FacDum {
        locus: Locus,
        corpus: Box<Stmt>,
        cond: Expr,
    },

    #[serde(rename = "Iteratio")]
    Iteratio {
        locus: Locus,
        species: String,
        binding: String,
        iter: Expr,
        corpus: Box<Stmt>,
        asynca: bool,
    },

    #[serde(rename = "Elige")]
    Elige {
        locus: Locus,
        discrim: Expr,
        casus: Vec<EligeCasus>,
        #[serde(rename = "default_")]
        default: Option<Box<Stmt>>,
    },

    #[serde(rename = "Discerne")]
    Discerne {
        locus: Locus,
        discrim: Vec<Expr>,
        casus: Vec<DiscerneCasus>,
    },

    #[serde(rename = "Custodi")]
    Custodi {
        locus: Locus,
        clausulae: Vec<CustodiClausula>,
    },

    #[serde(rename = "Tempta")]
    Tempta {
        locus: Locus,
        corpus: Box<Stmt>,
        cape: Option<CapeClausula>,
        demum: Option<Box<Stmt>>,
    },

    #[serde(rename = "Redde")]
    Redde {
        locus: Locus,
        valor: Option<Expr>,
    },

    #[serde(rename = "Iace")]
    Iace {
        locus: Locus,
        arg: Expr,
        fatale: bool,
    },

    #[serde(rename = "Scribe")]
    Scribe {
        locus: Locus,
        gradus: String,
        args: Vec<Expr>,
    },

    #[serde(rename = "Adfirma")]
    Adfirma {
        locus: Locus,
        cond: Expr,
        msg: Option<Expr>,
    },

    #[serde(rename = "Rumpe")]
    Rumpe { locus: Locus },

    #[serde(rename = "Perge")]
    Perge { locus: Locus },

    #[serde(rename = "Incipit")]
    Incipit {
        locus: Locus,
        corpus: Box<Stmt>,
        asynca: bool,
    },

    #[serde(rename = "Probandum")]
    Probandum {
        locus: Locus,
        nomen: String,
        corpus: Vec<Stmt>,
    },

    #[serde(rename = "Proba")]
    Proba {
        locus: Locus,
        nomen: String,
        corpus: Box<Stmt>,
    },
}

impl Stmt {
    pub fn locus(&self) -> Locus {
        match self {
            Stmt::Massa { locus, .. } => *locus,
            Stmt::Expressia { locus, .. } => *locus,
            Stmt::Varia { locus, .. } => *locus,
            Stmt::Functio { locus, .. } => *locus,
            Stmt::Genus { locus, .. } => *locus,
            Stmt::Pactum { locus, .. } => *locus,
            Stmt::Ordo { locus, .. } => *locus,
            Stmt::Discretio { locus, .. } => *locus,
            Stmt::TypusAlias { locus, .. } => *locus,
            Stmt::In { locus, .. } => *locus,
            Stmt::Importa { locus, .. } => *locus,
            Stmt::Si { locus, .. } => *locus,
            Stmt::Dum { locus, .. } => *locus,
            Stmt::FacDum { locus, .. } => *locus,
            Stmt::Iteratio { locus, .. } => *locus,
            Stmt::Elige { locus, .. } => *locus,
            Stmt::Discerne { locus, .. } => *locus,
            Stmt::Custodi { locus, .. } => *locus,
            Stmt::Tempta { locus, .. } => *locus,
            Stmt::Redde { locus, .. } => *locus,
            Stmt::Iace { locus, .. } => *locus,
            Stmt::Scribe { locus, .. } => *locus,
            Stmt::Adfirma { locus, .. } => *locus,
            Stmt::Rumpe { locus } => *locus,
            Stmt::Perge { locus } => *locus,
            Stmt::Incipit { locus, .. } => *locus,
            Stmt::Probandum { locus, .. } => *locus,
            Stmt::Proba { locus, .. } => *locus,
        }
    }
}

/// Function parameter.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Param {
    pub locus: Locus,
    pub nomen: String,
    pub typus: Option<Typus>,
    #[serde(rename = "default_")]
    pub default: Option<Expr>,
    pub rest: bool,
    /// Ownership annotation: "ex" (consume), "de" (borrow), "in" (mutate)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ownership: Option<String>,
}

/// Field declaration in genus.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CampusDecl {
    pub locus: Locus,
    pub nomen: String,
    pub typus: Option<Typus>,
    pub valor: Option<Expr>,
    pub visibilitas: String,
}

/// Protocol method declaration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PactumMethodus {
    pub locus: Locus,
    pub nomen: String,
    pub params: Vec<Param>,
    #[serde(rename = "typusReditus")]
    pub typus_reditus: Option<Typus>,
    pub asynca: bool,
}

/// Enum member.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrdoMembrum {
    pub locus: Locus,
    pub nomen: String,
    pub valor: Option<String>,
}

/// Discriminated union variant.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VariansDecl {
    pub locus: Locus,
    pub nomen: String,
    pub campi: Vec<VariansCampus>,
}

/// Field in a variant.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VariansCampus {
    pub nomen: String,
    pub typus: Typus,
}

/// Import specifier.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ImportSpec {
    pub locus: Locus,
    pub imported: String,
    pub local: String,
}

/// Case in elige (switch).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EligeCasus {
    pub locus: Locus,
    pub cond: Expr,
    pub corpus: Box<Stmt>,
}

/// Case in discerne (pattern match).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiscerneCasus {
    pub locus: Locus,
    pub patterns: Vec<VariansPattern>,
    pub corpus: Box<Stmt>,
}

/// Pattern in discerne case.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VariansPattern {
    pub locus: Locus,
    pub variant: String,
    pub bindings: Vec<String>,
    pub alias: Option<String>,
    pub wildcard: bool,
}

/// Guard clause.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CustodiClausula {
    pub locus: Locus,
    pub cond: Expr,
    pub corpus: Box<Stmt>,
}

/// Catch clause.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CapeClausula {
    pub locus: Locus,
    pub param: String,
    pub corpus: Box<Stmt>,
}

/// Top-level compilation unit.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Modulus {
    pub locus: Locus,
    pub corpus: Vec<Stmt>,
}
